### PR TITLE
[MM-55004] Add warning log message when app run as root

### LINKED ISSUE #24996: Log warning if Mattermost is run as root
It’s generally not advised to run Mattermost as `root`. This ticket is about adding a warning log message on startup if the mattermost binary is run as `root`.

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-55004

### PR DESCRIPTION
**Pull request by Paul-vrn**
#### Summary

Add a warning log message when Mattermost app is running with root user

#### Ticket Link

Fixes #24996 
JIRA: https://mattermost.atlassian.net/browse/MM-55004

#### Release Note

```release-note
Add warning log message when app run as root 
```

### CODE REVIEW

File: server/channels/app/server.go [hunk 1]
```diff
@@ -996,6 +996,11 @@ func (s *Server) Start() error {
 	logListeningPort := fmt.Sprintf("Server is listening on %v", listener.Addr().String())
 	mlog.Info(logListeningPort, mlog.String("address", listener.Addr().String()))
 
+	// Check if the command is called by root user
+	if os.Geteuid() == 0 {
+		mlog.Warn("Running as root is not recommended")
+	}
```
Comment (hanzei) [inline server/channels/app/server.go:1002]: Let's move that code closer to other sanity checks. Next to https://github.com/mattermost/mattermost/blob/37da0531919032722dfc8036b293bd1bdc900a7b/server/channels/app/server.go#L415-L417 would be a good place.

File: server/channels/app/server.go [hunk 2]
```diff
@@ -996,6 +996,11 @@ func (s *Server) Start() error {
 	logListeningPort := fmt.Sprintf("Server is listening on %v", listener.Addr().String())
 	mlog.Info(logListeningPort, mlog.String("address", listener.Addr().String()))
 
+	// Check if the command is called by root user
+	if os.Geteuid() == 0 {
+		mlog.Warn("Running as root is not recommended")
```
Comment (hanzei) [inline server/channels/app/server.go:1001]:
```suggestion
		mlog.Warn("Running Mattermost as root is not recommended. Please use a separate user")
```

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for the contribution @Paul-vrn Let's ensure the warning is placed close to other similar messanges :+1:

**Review by hanzei — APPROVED** [review]

**Comment by streamer45** [comment]
@hanzei I wonder if we should add the check at the command level. You can technically still run the binary as root without getting a warning if you run any other subcommand. I'd argue that would equally be a bad idea in most cases so maybe we could just move the check to the entry point `cmd/mattermost/main.go`.

**Comment by Paul-vrn** [comment]
> @hanzei I wonder if we should add the check at the command level. You can technically still run the binary as root without getting a warning if you run any other subcommand. I'd argue that would equally be a bad idea in most cases so maybe we could just move the check to the entry point `cmd/mattermost/main.go`.

Yeah, that's what I implemented in the beginning. But I saw most of the log messages being generated in the `app/server.go`, so I changed my mind. 😄
I'll make the change depending on what Hanzei suggests!

**Comment by hanzei** [comment]
Good point. Most commands are not safe to run as root. Let's move the check.

File: server/channels/app/server.go [hunk 3]
```diff
@@ -411,7 +411,7 @@ func NewServer(options ...Option) (*Server, error) {
 			http.Redirect(w, r, r.URL.String(), http.StatusFound)
 		})
 	}
-
+
```
Comment (hanzei) [inline server/channels/app/server.go:414]:
```suggestion

```

Comment (hanzei) [reply server/channels/app/server.go:414]: Please remove the unrelated change.

File: server/cmd/mattermost/main.go [hunk 4]
```diff
@@ -17,6 +19,12 @@ import (
 )
 
 func main() {
+
+	// Check if the command is called by root user
+	if os.Geteuid() == 0 {
+		mlog.Warn("Running Mattermost as root is not recommended. Please use a separate user")
+	}
```
Comment (hanzei) [inline server/cmd/mattermost/main.go:26]: While the code works correctly, I think it should be placed in `root.go` instead to keep `main.go` clean. You can use the `PersistentPreRun` field of `RootCmd`.

**Review by hanzei — CHANGES_REQUESTED** [review]

**Comment by streamer45** [comment]
@Paul-vrn Looks like our linter is not happy with some minor formatting. Could you run `gofmt -s -w cmd/mattermost/commands/root.go` and commit?

**Comment by Paul-vrn** [comment]
> gofmt -s -w cmd/mattermost/commands/root.go

ah yess didn't see, my bad, next time I'll run this command thx!

**Review by streamer45 — APPROVED** [review]
Looks good! Thank you for the contribution @Paul-vrn! :tada:

File: server/channels/app/server.go [hunk 5]
```diff
@@ -411,7 +411,6 @@ func NewServer(options ...Option) (*Server, error) {
 			http.Redirect(w, r, r.URL.String(), http.StatusFound)
 		})
 	}
-
 	if _, err = url.ParseRequestURI(*s.platform.Config().ServiceSettings.SiteURL); err != nil {
```
Comment (hanzei) [inline server/channels/app/server.go:414]: Nit: Would you mind removing the unrelated change?

File: server/cmd/mattermost/main.go [hunk 6]
```diff
@@ -7,6 +7,7 @@ import (
 	"os"
 
 	"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands"
+
```
Comment (hanzei) [inline server/cmd/mattermost/main.go:10]: Same here

**Review by hanzei — APPROVED** [review]

**Comment by hanzei** [comment]
Thank you very much for your contribution and for bearing with me while we figure out the best solution @Paul-vrn !

**Review by yasserfaraazkhan — APPROVED** [review]
🎉 
image

**Pull request merged**